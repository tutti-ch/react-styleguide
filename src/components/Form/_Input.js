import React, { Component } from "react"
import PropTypes from "prop-types"
import { filterProps } from "../../helpers/functions"
import WithWrapper from "./_WithWrapper"

export class Input extends Component {
  static defaultProps = {
    value: "",
    type: "text",
  }

  static propTypes = {
    /**
     * The initial value.
     */
    value: PropTypes.string,

    /**
     * The input type.
     */
    type: PropTypes.string,

    /**
     * The input name.
     */
    name: PropTypes.string,
  }

  constructor(props) {
    super(props)

    this.state = {
      value: props.value,
    }

    this.handleChange = this.handleChange.bind(this)
  }

  /**
   * Sync the value.
   */
  componentWillReceiveProps(nextProps) {
    if (nextProps.value !== this.props.value) {
      this.setState({ value: nextProps.value })
    }
  }

  /**
   * Handle the change event.
   */
  handleChange(event) {
    this.setState({ value: event.target.value }, () => {
      if (typeof this.props.onChange === "function") {
        const { name, value } = this.props
        this.props.onChange(this.state.value, { name, initialValue: value })
      }
    })
  }

  render() {
    const { value } = this.state
    const { type, name } = this.props
    const props = filterProps(Input.propTypes, this.props)

    return <input
      {...props}
      name={name}
      type={type}
      value={value}
      onChange={this.handleChange}
    />
  }
}

export default WithWrapper(Input)
