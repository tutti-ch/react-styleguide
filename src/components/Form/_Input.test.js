/* global describe, test, expect, jest */
import React from "react";
import { Input } from "./_Input";
import { shallow, mount } from "enzyme";

describe("(Component) Input", () => {
  test("[componentWillReceiveProps] should sync the state with props", () => {
    const comp = shallow(<Input />);
    const inst = comp.instance();
    inst.componentWillReceiveProps({ value: "Frocello" });
    expect(comp.state("value")).toBe("Frocello");
  });

  test("[handleChange] should set the value and call the onChange from parent", () => {
    const comp = shallow(<Input />);
    const inst = comp.instance();
    inst.handleChange({ target: { value: "Moreno" } });
    expect(comp.state("value")).toBe("Moreno");
  });

  test("[handleBlur] should notify the parent and call attached onBlur", () => {
    const onBlur = jest.fn();
    const onChange = jest.fn();
    const comp = shallow(<Input onBlur={onBlur} onChange={onChange} />);
    const inst = comp.instance();
    const event = {};
    inst.handleBlur(event);
    expect(onBlur).toHaveBeenCalledWith(event);
    expect(onChange).not.toHaveBeenCalled(); // Because nothing changed.
  });

  test("[handleKeyDown] should notify the parent and call attached onKeyDown", () => {
    const onKeyDown = jest.fn();
    const onChange = jest.fn();
    const comp = shallow(<Input onKeyDown={onKeyDown} onChange={onChange} />);
    const inst = comp.instance();
    inst.handleChange({ target: { value: "abc" } });
    inst.handleKeyDown({ key: "Enter" });
    expect(onKeyDown).toHaveBeenCalledWith({ key: "Enter" });
    expect(onChange).toHaveBeenCalled();

    comp.setProps({ onKeyDown: undefined });
    inst.handleChange({ target: { value: "def" } });
    inst.handleKeyDown({ key: "ArrowDown" });
    expect(onKeyDown).toHaveBeenCalledTimes(1); // Because we have set to undefined
    expect(onChange).toHaveBeenCalledTimes(1); // Because they key up was not an enter ðŸ˜€
  });

  test("Should call notify when new props are received", () => {
    const notify = jest.fn();
    const comp = shallow(<Input />);
    const inst = comp.instance();
    inst.notify = notify;
    comp.setProps({ value: "newValue" });
    expect(notify).toHaveBeenCalled();
  });

  test("snapshot", () => {
    const comp = mount(<Input value="Life is good" type="text" />);
    expect(comp).toMatchSnapshot();
  });
});
