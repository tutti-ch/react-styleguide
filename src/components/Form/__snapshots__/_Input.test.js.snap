// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`(Component) Input snapshot 1`] = `
<Input
  type="text"
  value="Life is good"
>
  <input
    name={undefined}
    onChange={[Function]}
    type="text"
    value="Life is good"
  />
</Input>
`;
